app/views/concerns/facet_presenter.rb

```diff
@@ -0,0 +1,124 @@
+module FacetPresenter
+  extend ActiveSupport::Concern
+
+  protected
+
+  def facet_item_url(facet, item)
+    if facet_in_params?(facet, item)
+      search_action_url(remove_facet_params(facet, item, params))
+    else
+      search_action_url(add_facet_params_and_redirect(facet, item))
+    end
+  end
+
+  def facet_display(facet, children = false)
+    facet_config = blacklight_config.facet_fields[facet.name]
+    if facet_config.colour
+      colour_facet(facet)
+    elsif facet_config.range
+      range_facet(facet)
+    elsif facet_config.hierarchical && !facet_config.parent
+      hierarchical_facet(facet)
+    else
+      simple_facet(facet)
+    end
+  end
+
+  def basic_facet(facet, type = :simple)
+    facet_config = blacklight_config.facet_fields[facet.name]
+    {
+      title: facet_map(facet.name),
+      select_one: facet_config.single,
+      items: facet.items[0..3].map { |item| send(:"#{type}_facet_item", facet, item) },
+      extra_items: facet.items.size <= 4 ? nil : {
+        items: facet.items[4..-1].map { |item| send(:"#{type}_facet_item", facet, item) }
+      }
+    }.merge(type => true)
+  end
+
+  def basic_facet_item(facet, item)
+    {
+      url: facet_item_url(facet.name, item),
+      text: facet_map(facet.name, item.value),
+      num_results: number_with_delimiter(item.hits),
+      is_checked: facet_in_params?(facet.name, item)
+    }
+  end
+
+  def simple_facet(facet)
+    basic_facet(facet, type = :simple)
+  end
+
+  def simple_facet_item(facet, item)
+    basic_facet_item(facet, item)
+  end
+
+  def colour_facet(facet)
+    basic_facet(facet, type = :colour)
+  end
+
+  def colour_facet_item(facet, item)
+    basic_facet_item(facet, item).tap do |basic|
+      basic.delete(:text)
+      basic[:hex] = item.value
+    end
+  end
+
+  def hierarchical_facet(facet)
+    basic_facet(facet, type = :hierarchical)
+  end
+
+  def hierarchical_facet_item(facet, item)
+    child_facets = hierarchical_facet_item_children(facet, item)
+
+    {
+      has_subfilters: child_facets.present?,
+      filters: child_facets.map { |child| facet_display(child, true) }
+    }.merge(simple_facet_item(facet, item))
+  end
+
+  def hierarchical_facet_item_children(facet, item)
```

Comment:
Assignment Branch Condition size for hierarchical_facet_item_children is too high. [16.16/15]